lib/block_view_stake.go

```diff
@@ -0,0 +1,2496 @@
+	lockedStakeEntries, err := bav.GetLockedStakeEntriesInRange(
+		validatorEntry.ValidatorPKID, transactorPKIDEntry.PKID, metadata.StartEpochNumber, metadata.EndEpochNumber,
+	)
+	existsLockedStakeEntries := false
+	for _, lockedStakeEntry := range lockedStakeEntries {
+		if lockedStakeEntry != nil && !lockedStakeEntry.isDeleted {
+			existsLockedStakeEntries = true
+			break
+		}
+	}
+	if !existsLockedStakeEntries {
+		return RuleErrorInvalidUnlockStakeNoUnlockableStakeFound
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) GetStakeEntry(validatorPKID *PKID, stakerPKID *PKID) (*StakeEntry, error) {
+	// First, check the UtxoView.
+	stakeMapKey := StakeMapKey{ValidatorPKID: *validatorPKID, StakerPKID: *stakerPKID}
+	if stakeEntry, exists := bav.StakeMapKeyToStakeEntry[stakeMapKey]; exists {
+		// If StakeEntry.isDeleted, return nil.
+		if stakeEntry.isDeleted {
+			return nil, nil
+		}
+		return stakeEntry, nil
+	}
+	// Then, check the database.
+	stakeEntry, err := DBGetStakeEntry(bav.Handle, bav.Snapshot, validatorPKID, stakerPKID)
+	if err != nil {
+		return nil, err
+	}
+	if stakeEntry != nil {
+		// Cache the StakeEntry in the UtxoView if exists.
+		bav._setStakeEntryMappings(stakeEntry)
+	}
+	return stakeEntry, nil
+}
+
+func (bav *UtxoView) GetLockedStakeEntry(
+	validatorPKID *PKID,
+	stakerPKID *PKID,
+	lockedAtEpochNumber uint64,
+) (*LockedStakeEntry, error) {
+	// First, check the UtxoView.
+	lockedStakeMapKey := LockedStakeMapKey{
+		ValidatorPKID:       *validatorPKID,
+		StakerPKID:          *stakerPKID,
+		LockedAtEpochNumber: lockedAtEpochNumber,
+	}
+	if lockedStakeEntry, exists := bav.LockedStakeMapKeyToLockedStakeEntry[lockedStakeMapKey]; exists {
+		// If LockedStakeEntry.isDeleted, return nil.
+		if lockedStakeEntry.isDeleted {
+			return nil, nil
+		}
+		return lockedStakeEntry, nil
+	}
+	// Then, check the database.
+	lockedStakeEntry, err := DBGetLockedStakeEntry(bav.Handle, bav.Snapshot, validatorPKID, stakerPKID, lockedAtEpochNumber)
+	if err != nil {
+		return nil, err
+	}
+	if lockedStakeEntry != nil {
+		// Cache the LockedStakeEntry in the UtxoView if exists.
+		bav._setLockedStakeEntryMappings(lockedStakeEntry)
+	}
+	return lockedStakeEntry, nil
+}
+
+func (bav *UtxoView) GetLockedStakeEntriesInRange(
+	validatorPKID *PKID,
+	stakerPKID *PKID,
+	startEpochNumber uint64,
+	endEpochNumber uint64,
+) ([]*LockedStakeEntry, error) {
+	// Store matching LockedStakeEntries in a set to prevent
+	// returning duplicates between the db and UtxoView.
+	lockedStakeEntriesSet := NewSet([]*LockedStakeEntry{})
+
+	// First, pull matching LockedStakeEntries from the db.
+	dbLockedStakeEntries, err := DBGetLockedStakeEntriesInRange(
+		bav.Handle, bav.Snapshot, validatorPKID, stakerPKID, startEpochNumber, endEpochNumber,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "UtxoView.GetLockedStakeEntriesInRange: ")
+	}
+	for _, lockedStakeEntry := range dbLockedStakeEntries {
+		lockedStakeEntriesSet.Add(lockedStakeEntry)
```

Comment:
hmm upon more thought, I'm a little concerned about using sets here. I'm worried that the comparable stuff won't work w/ pointers.

Reply:
Okay so I wrote a quick test and it fails. So we will want to do something a bit different. I'd recommend creating a map of LockedStakeMapKey to LockedStakeEntry like we have in the utxo view. Here's the test for reference:

```
func TestStructSetAdd(t *testing.T) {
	lockedStakeEntry := &LockedStakeEntry{
		LockedStakeID:       &ZeroBlockHash,
		StakerPKID:          &ZeroPKID,
		ValidatorPKID:       &ZeroPKID,
		LockedAmountNanos:   uint256.NewInt(),
		LockedAtEpochNumber: 100,
	}
	set := NewSet([]*LockedStakeEntry{lockedStakeEntry})

	set.Add(lockedStakeEntry.Copy())

	require.Equal(t, set.Size(), 1)
}
```